shell-integration/fish/vendor_conf.d/kitty-shell-integration.fish

```diff
@@ -20,66 +22,177 @@ function __ksi_schedule --on-event fish_prompt -d "Setup kitty integration after
     set --local _ksi (string split " " -- "$KITTY_SHELL_INTEGRATION")
     set --erase KITTY_SHELL_INTEGRATION KITTY_FISH_XDG_DATA_DIR
 
-    function __ksi_osc
-        printf "\e]%s\a" "$argv[1]"
-    end
-
+    # Enable cursor shape changes for default mode and vi mode
+    #
+    # For vi mode, the prompt marking feature needs to be enabled,
+    # otherwise it falls back to fish's default fish_vi_cursor.
+    #
+    # Custom cursor shapes can be configured via global variables.
+    #
+    # For normal mode prompt:
+    #     kitty_fish_prompt_cursor
+    # For programs running in fish:
+    #     kitty_fish_external_cursor
+    # For vi modes:
+    #     fish_{default,insert,replace_one,visual}_cursor
+    # 
+    # Cursor shape values:
+    #     block, underscore (underline), line (bar, beam)
+    #
+    # To make the cursor blink, add `blink` at the end.
+    #
+    # Example:
+    #     set -g kitty_fish_prompt_cursor line blink
+    #
     if not contains "no-cursor" $_ksi
         and not functions -q __ksi_set_cursor
 
-        function __ksi_set_cursor --on-variable fish_key_bindings -d "Set cursor shape for fish default mode"
-            if test "$fish_key_bindings" = fish_default_key_bindings
-                not functions -q __ksi_bar_cursor __ksi_block_cursor || return
+        # Set cursor shape escape sequence `ESC [ Ps SPACE q`
+        # Ps:
+        # blinking: block 1, underline 3, bar 5
+        # steady:   block 2, underline 4, bar 6
+        function __ksi_init_cursor_var -d "Initialize cursor shape CSI escape sequence"
+            set --local varname kitty_fish_$argv[1]_cursor
+            set --local val 0
+            if contains $argv[1] default insert replace_one visual
+                and test -n "fish_$argv[1]_cursor"
+                set varname fish_$argv[1]_cursor
+            end
+            if test -n "$$varname"
+                set --local shape (string split " " -- "$$varname")
+                switch $shape[1]
+                    case block
+                        set val 2
+                    case underline underscore
+                        set val 4
+                    case bar beam line
```

Comment:
Why does this have multiple names? Fish only knows "block", "underscore" and "line".

This means people now go setting their var to "beam" and then come reporting bugs to us if it doesn't work elsewhere. E.g. if they read a blog post of someone who uses kitty.

Reply:
The "bar" in kitty's config is called "beam". I'd like to make either name work, sorry, maybe it should be strictly fish naming. These definitely need to be removed.